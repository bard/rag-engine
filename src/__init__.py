from . import ingest
from . import query
