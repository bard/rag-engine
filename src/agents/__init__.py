from .document_processor import DocumentProcessor
from .vector_store import VectorStore

__all__ = ["DocumentProcessor", "VectorStore"]
