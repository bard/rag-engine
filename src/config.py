from langchain_core.runnables.config import RunnableConfig
from pydantic import BaseModel, Field
from typing import Self, Union, Literal


class PineconeVectorStoreBackend(BaseModel):
    type: Literal["pinecone"]
    index_name: str
    api_key: str
    embeddings_type: Literal["openai"]
    score_threshold: float = 0.1


class MockVectorStoreBackend(BaseModel):
    type: Literal["mock"]
    score_threshold: float = 0.1


class ChromaVectorStoreBackend(BaseModel):
    type: Literal["chroma"]
    path: str
    collection_name: str
    embeddings_type: Literal["local-minilm", "openai"]
    score_threshold: float = 0.1


type VectorStoreBackend = Union[
    PineconeVectorStoreBackend, MockVectorStoreBackend, ChromaVectorStoreBackend
]


class OpenaiLlmBackend(BaseModel):
    type: Literal["openai"]
    model: str
    api_key: str


class AnthropicLlmBackend(BaseModel):
    type: Literal["anthropic"]
    model: str
    api_key: str


class DbBackend(BaseModel):
    url: str


type LlmBackend = Union[OpenaiLlmBackend, AnthropicLlmBackend]


class AgentConfig(BaseModel):
    db: DbBackend
    llm: LlmBackend = Field(discriminator="type")
    vector_store: VectorStoreBackend = Field(discriminator="type")

    @classmethod
    def from_runnable_config(cls, config: RunnableConfig) -> Self:
        """Validate runnable config"""
        configurable = config.get("configurable")
        assert configurable is not None
        return cls(**configurable)
