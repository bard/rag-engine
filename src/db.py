import uuid
from typing import List
from datetime import datetime, timezone
from sqlalchemy.orm import Mapped, mapped_column, relationship, DeclarativeBase
from sqlalchemy import inspect, DateTime, ForeignKey


class SqlAlchemyBase(DeclarativeBase):
    pass


class SqlTopic(SqlAlchemyBase):
    """SqlAlchemy model for stored topic"""

    __tablename__ = "topics"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        default=lambda: datetime.now(timezone.utc), nullable=False
    )

    # Relationship to KnowledgeBase
    knowledge_documents: Mapped[List["SqlKnowledgeBaseDocument"]] = relationship(
        back_populates="topic", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"SqlTopic(id={self.id}, name={self.name})"


class SqlKnowledgeBaseDocument(SqlAlchemyBase):
    """SqlAlchemy model for stored document"""

    __tablename__ = "documents"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: str(uuid.uuid4()))
    data: Mapped[str] = mapped_column(nullable=False)
    content: Mapped[str] = mapped_column(nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    topic_id: Mapped[str] = mapped_column(
        ForeignKey("topics.id"),
        # allow NULL to signify "default topic"
        nullable=True,
    )
    topic: Mapped[SqlTopic] = relationship(back_populates="knowledge_documents")

    def __repr__(self) -> str:
        return f"SqlKnowledgeBaseDocument(id={self.id}, data={self.data}, content={self.content}"

    def to_dict(self):
        return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}
