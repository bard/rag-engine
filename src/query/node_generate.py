import pprint
from typing import TypedDict
from langchain.schema import BaseMessage, Document, HumanMessage, SystemMessage
from langchain_core.runnables import RunnableConfig

from ..config import AgentConfig
from ..services import get_llm
from .state import AgentState


class GenerateStateUpdate(TypedDict):
    messages: list[BaseMessage]


def generate(state: AgentState, config: RunnableConfig) -> GenerateStateUpdate:
    """Generate answer."""
    c = AgentConfig.from_runnable_config(config)
    llm = get_llm(c)

    documents = state.get("documents")
    assert documents is not None
    query = state.get("query")
    assert query is not None
    weather_info = state.get("weather_info")
    location = state.get("location")

    docs_content = "\n\n".join(doc.page_content for doc in documents)

    system_message_content = (
        "You are an assistant for question-answering tasks. "
        "Use the following pieces of retrieved context to answer "
        "the question. If you don't know the answer, say that you "
        "don't know. Use three sentences maximum and keep the "
        "answer concise."
        "\n\n"
    )
    if weather_info is not None and location is not None:
        system_message_content += (
            f"Current weather information for {location}: {weather_info}\n\n"
        )

    system_message_content += f"{docs_content}\n\n"

    prompt = [SystemMessage(system_message_content), HumanMessage(query)]

    response = llm.invoke(prompt)
    return {"messages": [response]}
