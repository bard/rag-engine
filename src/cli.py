import click
import logging
from os import path
from langchain_core.messages import HumanMessage, AIMessage
from sqlalchemy.orm import Session

from src import services, workflow_query, workflow_ingest, db
from src.config import Config
from src.db import SqlTopic


CONFIG = Config.from_env()


@click.group()
@click.option(
    "--log",
    help="Set logging level",
    default="INFO",
    type=click.Choice(
        ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"], case_sensitive=False
    ),
)
def cli(log):
    logging.basicConfig(level=getattr(logging, log.upper()))


@click.command(name="create_topic")
@click.option("--id", required=False, help="Optional topic ID")
@click.option("--name", required=True, help="Topic name")
def cmd_create_topic(id: str | None, name: str):
    db = services.get_db(CONFIG)
    with Session(db) as session:
        topic = SqlTopic(id=id, name=name)
        session.add(topic)
        session.commit()
        session.refresh(topic)

    click.echo()
    click.echo(f"Created topic '{name}' with ID: {topic.id}")
    click.echo()


@click.command(name="ingest")
@click.argument("path_or_url")
@click.option("--topic_id", required=False, help="Optional topic ID")
def cmd_ingest(path_or_url: str, topic_id: str | None):
    if path_or_url.startswith(("http://", "https://", "file://", "data://")):
        input_url = path_or_url
    else:
        abs_path = path.abspath(path_or_url)
        input_url = f"file://{abs_path}"
    initial_agent_state = workflow_ingest.AgentState(
        url=input_url, source_content=None, extracted_data=[], topic_id=topic_id
    )

    workflow_ingest.get_graph().invoke(initial_agent_state, CONFIG.to_runnable_config())

    click.echo()
    click.echo("Data ingested successfully")
    click.echo()


@click.command(name="list_topics")
def cmd_list_topics():
    """List all available topics."""
    db_engine = services.get_db(CONFIG)
    with Session(db_engine) as session:
        topics = session.query(SqlTopic).order_by(SqlTopic.name).all()

    click.echo()
    if topics:
        click.echo("Available topics:")
        for topic in topics:
            click.echo(f"  {topic.id}: {topic.name}")
    else:
        click.echo("No topics found")
    click.echo()


@click.command(name="info")
def cmd_info():
    """Display information about the database and vector store."""
    # Get database stats
    db_engine = services.get_db(CONFIG)
    with Session(db_engine) as session:
        topic_count = session.query(SqlTopic).count()
        doc_count = session.query(db.SqlKnowledgeBaseDocument).count()

    # Get sanitized DB URL (remove credentials)
    db_url = str(CONFIG.db.url)
    if "@" in db_url:
        db_url = db_url.split("@")[1]
        db_url = f"***:***@{db_url}"

    click.echo()
    click.echo("Database Information:")
    click.echo(f"  URL: {db_url}")
    click.echo(f"  Topics: {topic_count}")
    click.echo(f"  Documents: {doc_count}")
    click.echo()


@click.command(name="initdb")
def cmd_initdb():
    engine = services.get_db(CONFIG)

    db.SqlAlchemyBase.metadata.create_all(engine)

    click.echo()
    click.echo("Database initialized successfully")
    click.echo()


@click.command(name="query")
@click.argument("user_query")
@click.option("--topic_id", required=False, help="Optional topic ID")
def cmd_query(user_query: str, topic_id: str | None):
    initial_agent_state = workflow_query.AgentState(
        messages=[HumanMessage(content=user_query)],
        retrieved_knowledge=[],
        query=None,
        topic_id=topic_id,
        external_knowledge_sources=[],
    )

    response = workflow_query.get_graph().invoke(
        initial_agent_state, CONFIG.to_runnable_config()
    )

    ai_messages = [msg for msg in response["messages"] if isinstance(msg, AIMessage)]
    click.echo()
    if ai_messages:
        click.echo(ai_messages[-1].content)
    else:
        click.echo("No AI response found")
    click.echo()


if __name__ == "__main__":
    cli.add_command(cmd_create_topic)
    cli.add_command(cmd_ingest)
    cli.add_command(cmd_info)
    cli.add_command(cmd_initdb)
    cli.add_command(cmd_list_topics)
    cli.add_command(cmd_query)
    cli()
