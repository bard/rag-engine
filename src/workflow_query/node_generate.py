import pprint
from typing import TypedDict
from langchain.schema import BaseMessage, HumanMessage, SystemMessage
from langchain_core.runnables import RunnableConfig
from sqlalchemy.orm import Session

from ..config import Config
from .. import services
from ..db import SqlTopic
from .state import AgentState


class GenerateStateUpdate(TypedDict):
    messages: list[BaseMessage]


def generate(state: AgentState, config: RunnableConfig) -> GenerateStateUpdate:
    conf = Config.from_runnable_config(config)

    query = state["query"]
    assert query is not None

    system_message_content = (
        "You are an assistant for question-answering tasks. "
        "Use the following pieces of retrieved context to answer "
        "If you don't know the answer, say that you "
        "don't know. Use three sentences maximum and keep the "
        "answer concise."
        "\n\n"
    )

    if state["topic_id"]:
        db = services.get_db(conf)
        with Session(db) as session:
            topic = (
                session.query(SqlTopic).filter(SqlTopic.id == state["topic_id"]).first()
            )
            if topic:
                system_message_content += f"Topic: {topic}\n\n"

    system_message_content += "\n\n".join(
        doc.page_content for doc in state["retrieved_knowledge"]
    )

    prompt = [SystemMessage(system_message_content), HumanMessage(query)]

    response = services.get_llm(conf).invoke(prompt)

    return {"messages": [response]}
