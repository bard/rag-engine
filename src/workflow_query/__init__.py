from .state import AgentState
from .graph import get_graph
from .node_classify_query import classify_query
from .node_fetch_weather_info import fetch_weather_info
from .node_generate import generate
from .node_retrieve import retrieve
from .node_rerank__STUB import rerank
from .node_reformulate_query__STUB import reformulate_query

__all__ = [
    "AgentState",
    "get_graph",
    "classify_query",
    "fetch_weather_info",
    "generate",
    "reformulate_query",
    "retrieve",
    "rerank",
]
