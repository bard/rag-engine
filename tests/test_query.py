import pprint
from datetime import datetime, timezone
import pytest
from langchain.schema import HumanMessage, Document
from langchain_core.messages import HumanMessage
from sqlalchemy.orm import Session
from src.db import SqlTopic

from src.workflow_query import (
    AgentState,
    get_graph,
    retrieve_from_knowledge_base,
    reformulate_query,
    classify_query,
)
from src import services
from src.workflow_query.node_generate import generate


@pytest.mark.vcr
def test_classify_query_detects_weather_query(config):
    with Session(services.get_db(config)) as session:
        paris_topic = SqlTopic(name="Paris", created_at=datetime.now(timezone.utc))
        session.add(paris_topic)
        session.commit()
        session.refresh(paris_topic)

    agent_state = AgentState(
        messages=[HumanMessage(content="how is the weather?")],
        retrieved_knowledge=[],
        query=None,
        topic_id=paris_topic.id,
        external_knowledge_sources=[],
    )

    state_update = classify_query(agent_state, config.to_runnable_config())

    assert state_update == {
        "query": "how is the weather?",
        "external_knowledge_sources": [{"type": "weather", "location": "Paris"}],
    }


@pytest.mark.vcr
def test_classify_query_determines_query_not_to_be_about_weather_if_no_topic_provided(
    config,
):
    agent_state = AgentState(
        messages=[HumanMessage(content="how is the weather?")],
        retrieved_knowledge=[],
        query=None,
        topic_id=None,
        external_knowledge_sources=[],
    )

    state_update = classify_query(agent_state, config.to_runnable_config())

    assert state_update == {
        "query": "how is the weather?",
        "external_knowledge_sources": [],
    }


def test_retrieve_with_travel_knowledge_base(
    config, travel_knowledge_documents, snapshot
):
    vector_store = services.get_vector_store(config)
    vector_store.add_documents(travel_knowledge_documents)

    agent_state = AgentState(
        messages=[HumanMessage(content="what are some nice things to see in Paris?")],
        retrieved_knowledge=[],
        query="what are some nice things to see in Paris?",
        topic_id="Paris",
        external_knowledge_sources=[],
    )

    state_update = retrieve_from_knowledge_base(
        agent_state, config.to_runnable_config()
    )

    assert state_update["retrieved_knowledge"] == snapshot


def test_retrieve_with_insurance_knowledge_base(config):
    vector_store = services.get_vector_store(config)
    vector_store.add_texts(
        [
            """# Average expenditures 2012-2013
Year: 2012, Expenditure: $812.40, Change: 2.2%
Year: 2013, Expenditure: $841.06, Change: 3.5%
""",
            """# Average expenditures 2014-2015
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%
""",
            """# Average expenditures 2015-2016
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%
""",
        ],
        metadatas=[
            {"topic_id": "UNCATEGORIZED"},
            {"topic_id": "UNCATEGORIZED"},
            {"topic_id": "UNCATEGORIZED"},
        ],
    )
    agent_state = AgentState(
        messages=[
            HumanMessage(content="what is the expenditure in 2014?"),
        ],
        retrieved_knowledge=[],
        query="what is the expenditure in 2014?",
        topic_id=None,
        external_knowledge_sources=[],
    )

    state_update = retrieve_from_knowledge_base(
        agent_state, config.to_runnable_config()
    )

    documents_content = "\n".join(
        [doc.page_content for doc in state_update["retrieved_knowledge"]]
    )

    assert (
        documents_content
        == """# Average expenditures 2014-2015
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%

# Average expenditures 2015-2016
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%
"""
    )


@pytest.mark.vcr
def test_generate_from_knowledge_base(config, travel_knowledge_documents, snapshot):
    agent_state = AgentState(
        messages=[HumanMessage(content="what are some nice things to see in Paris?")],
        retrieved_knowledge=travel_knowledge_documents[0:1],
        query="what are some nice things to see in Paris?",
        topic_id=None,
        external_knowledge_sources=[],
    )

    state_update = generate(agent_state, config.to_runnable_config())

    assert state_update["messages"][-1].content == snapshot


@pytest.mark.vcr
def test_generate_from_external_service(config, snapshot):
    agent_state = AgentState(
        messages=[HumanMessage(content="what is the weather like in paris?")],
        retrieved_knowledge=[
            Document(
                page_content="Current weather information for paris: Temperature: -0.55°C, Feels like: -3.75°C, Humidity: 95%, Status: mist",
                id="external[weather]",
            )
        ],
        query="what is the weather like in paris?",
        topic_id="aaaa-bbbb",
        external_knowledge_sources=[],
    )

    state_update = generate(agent_state, config.to_runnable_config())

    assert state_update["messages"][-1].content == snapshot


@pytest.mark.vcr
def test_graph_with_weather_query(config, travel_knowledge_documents, snapshot):
    with Session(services.get_db(config)) as session:
        paris_topic = SqlTopic(name="Paris", created_at=datetime.now(timezone.utc))
        session.add(paris_topic)
        session.commit()
        session.refresh(paris_topic)

    services.get_vector_store(config).add_documents(travel_knowledge_documents)

    agent_state = AgentState(
        messages=[HumanMessage(content="what is the weather like?")],
        retrieved_knowledge=[],
        query=None,
        topic_id=paris_topic.id,
        external_knowledge_sources=[],
    )

    response = get_graph().invoke(agent_state, config.to_runnable_config())

    assert response.get("messages", [])[-1].content == snapshot


@pytest.mark.vcr
def test_graph_with_knowledge_query(config, travel_knowledge_documents, snapshot):
    with Session(services.get_db(config)) as session:
        paris_topic = SqlTopic(name="Paris", created_at=datetime.now(timezone.utc))
        session.add(paris_topic)
        session.commit()
        session.refresh(paris_topic)

    vector_store = services.get_vector_store(config)
    vector_store.add_documents(travel_knowledge_documents)
    agent_state = AgentState(
        messages=[HumanMessage(content="what are some nice things to see?")],
        retrieved_knowledge=[],
        query=None,
        topic_id=paris_topic.id,
        external_knowledge_sources=[],
    )

    response = get_graph().invoke(agent_state, config.to_runnable_config())

    assert response.get("messages", [])[-1].content == snapshot


@pytest.mark.parametrize(
    "user_query",
    [
        "What's the trend in auto insurance costs over the last 3 years?",
        # "Compare insurance costs between different regions",
        # "What factors most influence insurance costs?",
        # "Generate a summary of key findings from the data",
    ],
)
@pytest.mark.vcr
def test_graph_with_insurance_queries(
    config, user_query, insurance_data_documents, snapshot
):
    services.get_vector_store(config).add_documents(insurance_data_documents)

    agent_state = AgentState(
        messages=[HumanMessage(content=user_query)],
        retrieved_knowledge=[],
        query=None,
        topic_id=None,
        external_knowledge_sources=[],
    )

    response = get_graph().invoke(agent_state, config.to_runnable_config())

    assert response.get("messages", [])[-1].content == snapshot


@pytest.mark.xfail(reason="node not used")
@pytest.mark.vcr
def test_reformulate_query__STUB(config):
    agent_state = AgentState(
        messages=[
            HumanMessage(
                content="What's the trend in auto insurance costs from year 2012 to year 2014?",
            )
        ],
        retrieved_knowledge=[],
        query=None,
        topic_id=None,
        external_knowledge_sources=[],
    )

    state_update = reformulate_query(agent_state, config.to_runnable_config())

    assert state_update == {"query": "Auto insurance cost trend 2012-2014"}
