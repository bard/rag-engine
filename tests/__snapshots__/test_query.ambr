# serializer version: 1
# name: test_generate_from_external_service
  'The weather in Paris is currently misty with a temperature of -0.55°C, feeling like -3.75°C, and a humidity of 95%.'
# ---
# name: test_generate_from_knowledge_base
  'Some nice things to see in Paris include the Eiffel Tower for its panoramic views, the Louvre Museum for its art masterpieces like the Mona Lisa, and the Notre-Dame Cathedral for its Gothic architecture. Additionally, you can explore the Champs-Élysées leading to the Arc de Triomphe and visit Montmartre with the Basilica of the Sacré-Cœur.'
# ---
# name: test_graph_with_insurance_queries[What's the trend in auto insurance costs over the last 3 years?]
  'The average expenditures for auto insurance have been increasing over the last three years. From 2015 to 2017, the expenditures rose from $896.66 to $1008.35, with annual increases of 3.1%, 5.4%, and 6.7%, respectively. This indicates a trend of rising costs.'
# ---
# name: test_graph_with_knowledge_query
  'Some nice things to see in Paris include the Eiffel Tower, which offers panoramic views from its observation decks, the Louvre Museum with masterpieces like the Mona Lisa, and the Gothic architectural marvel of Notre-Dame Cathedral. The Champs-Élysées, leading to the Arc de Triomphe, and Montmartre with the Basilica of the Sacré-Cœur are also must-visit spots.'
# ---
# name: test_graph_with_weather_query
  'The weather in Paris is currently 6.75°C with a feels-like temperature of 2.7°C. Humidity is at 83%, and there is light rain.'
# ---
# name: test_retrieve_with_travel_knowledge_base
  list([
    Document(id='knowledge_base[0]', metadata={'topic_id': 'Paris'}, page_content="Paris, the 'City of Light,' boasts iconic landmarks such as the Eiffel Tower, offering panoramic views from its observation decks. The Louvre Museum, home to masterpieces like the Mona Lisa, attracts art enthusiasts worldwide. Notre-Dame Cathedral, a Gothic architectural marvel, stands on the Île de la Cité. The Champs-Élysées, lined with shops and cafes, leads to the Arc de Triomphe, honoring those who fought for France. Montmartre, with its artistic heritage, features the Basilica of the Sacré-Cœur atop its hill."),
  ])
# ---
